File: Estrutura_de_Dados/2021-05-20/Merge_Sort/typescript/mergeSort.ts
export function mergeSort(
  array: number[],
  left?: number,
  middle?: number,
  right?: number
) {
  left = left || 0;
  middle = middle || array.length / 2;
  right = right || array.length;

  // transfere os elementos entre left e right para um array auxiliar.
  let helper = [...array];

  let index = left;
  let indexJ = middle;
  let indexK = left;

  while (index <= middle && indexJ <= right) {
    if (helper[index] < helper[indexJ]) {
      array[indexK] = helper[index];
      index++;
    } else {
      array[indexK] = helper[indexJ];
      indexJ++;
    }
    indexK++;
  }

  // se a metade inicial não foi toda consumida, faz o append.
  while (index <= middle) {
    array[indexK] = helper[index];
    index++;
    indexK++;
  }

  // se a metade final não foi toda consumida, faz o append.
  while (indexJ <= right) {
    array[indexK] = helper[indexJ];
    indexJ++;
    indexK++;
  }
}

File: Estrutura_de_Dados/2021-05-20/Merge_Sort/python/mergeSort.py
def merge_sort(array: list, left: int = None, middle: int = None, right: int = None):
    left = left == None if left else 0
    middle = middle == None if middle else len(array) / 2
    right = right == None if right else len(array)

    helper = array.copy()

    index = int(left)
    indexJ = int(middle)
    indexK = int(left)

    while index < middle and indexJ < right:
        if helper[index] < helper[indexJ]:
            array[indexK] = helper[index]
            index += 1
        else:
            array[indexK] = helper[indexJ]
            indexJ += 1
        indexK += 1

    while index < middle:
        array[indexK] = helper[index]
        index += 1
        indexK += 1

    while indexJ < right:
        array[indexK] = helper[indexJ]
        indexJ += 1
        indexK += 1
